File: back_end/models/userModel.js
const mongoose = require('mongoose');
const validator = require('validator');
const bcrypt = require('bcryptjs');
const userSchema = new mongoose.Schema({
    name: {
        type: String,
        required: [true, 'A user must have a name'],
        trim: true,
        maxlength: [20, 'A user name must have less or equal then 20 characters'],
        minlength: [4, 'A user name must have more or equal then 4 characters'],
    },
    email: {
        type: String,
        required: [true, 'A user must have an email'],
        trim: true,
        unique: true,
        lowercase: true,
        validate: [validator.isEmail, 'Please provide a valid email']
    },
    password: {
        type: String,
        required: [true, 'A user must have a password'],
        trim: true,
        maxlength: [20, 'A user password must have less or equal then 20 characters'],
        minlength: [8, 'A user password must have more or equal then 8 characters'],
        select: false
    },
    // role: {
    //     type: String,
    //     enum: ['user', 'admin'],
    //     default: 'user'
    // },
    photo: {
        type: String,
        default: 'default.jpg'
    },
    phone: {
        type: String,
        // required: [true, 'A user must have a phone number'],
        trim: true,
        maxlength: [10, 'A user phone number must have less or equal then 10 characters'],
        minlength: [10, 'A user phone number must have more or equal then 10 characters'],
    },
    address: {
        type: String,
        // required: [true, 'A user must have an address'],
        trim: true,
        maxlength: [100, 'A user address must have less or equal then 100 characters'],
        minlength: [10, 'A user address must have more or equal then 10 characters'],
    },
    // cart: {
    //     type: Array,
    //     default: []
    // },
    // history: {
    //     type: Array,
    //     default: []
    // },
    passwordChangedAt: Date,
    // shop: {
    //     type: mongoose.Schema.Types.ObjectId,
    //     ref: 'Shop',
    // },
    stripeId: {
        type: String,
        default: '',
        required: [true, 'A user must have a stripeId'],
    },
    address: {
        type: Object,
        default: {
            detail: '',
            province_th: '',
            province_en: '',
            district_th: '',
            district_en: '',
            sub_district_th: '',
            sub_district_en: '',
            post_code: '',
        },
    }
},
    {
        timestamps: true,
        toJSON: { virtuals: true },
        toObject: { virtuals: true }
    },
    this.collection = 'users'
);

userSchema.virtual('shop', {
    ref: 'Shop',
    localField: '_id',
    foreignField: 'owner'
});

userSchema.pre('save', function (next) {
    // Only run this function if password was actually modified
    if (!this.isModified('password')) return next();

    // Hash the password with cost of 10
    this.password = bcrypt.hashSync(this.password, 10);
    next();
});

userSchema.methods.correctPassword = function (candidatePassword, userPassword) {
    return bcrypt.compareSync(candidatePassword, userPassword);
};

userSchema.methods.changedPasswordAfter = function (JWTTimestamp) {
    if (this.passwordChangedAt) {
        const changedTimestamp = parseInt(this.passwordChangedAt.getTime() / 1000, 10);
        // console.log(changedTimestamp, JWTTimestamp);
        return JWTTimestamp < changedTimestamp;
    }
    // False means NOT changed
    return false;
};

module.exports = mongoose.model('User', userSchema);

File: front_end/lib/src/models/user_model.dart
import 'shop_model.dart';

class User {
  final String photo;
  final String name;

  User({
    required this.photo,
    required this.name,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      photo: json['photo'] ??
          'https://getgoods.blob.core.windows.net/user-photos/default.png',
      name: json['name'] ?? '',
    );
  }

  factory User.empty() {
    return User(
      photo: 'https://getgoods.blob.core.windows.net/user-photos/default.png',
      name: '',
    );
  }
}

class UserDetail {
  final String id;
  final String photo;
  final String name;
  final String email;
  final String phone;
  //final String address;
  final Address address;
  final ShopDetail shop;

  UserDetail({
    required this.id,
    required this.photo,
    required this.name,
    required this.email,
    required this.phone,
    required this.address,
    required this.shop,
  });

  factory UserDetail.fromJson(Map<String, dynamic> json) {
    var shopJson = json['shop'];
    ShopDetail shop;
    if (shopJson == null || shopJson.isEmpty) {
      shop = ShopDetail.empty();
    } else if (shopJson is List) {
      shop = ShopDetail.fromJson(shopJson.first as Map<String, dynamic>);
    } else {
      shop = ShopDetail.fromJson(shopJson as Map<String, dynamic>);
    }
    return UserDetail(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      photo: json['photo'] ?? '',
      phone: json['phone'] ?? '',
      address: Address.fromJson(json['address'] ?? {}),
      shop: shop,
    );
  }

  factory UserDetail.empty() {
    return UserDetail(
      id: '',
      photo: '',
      name: '',
      email: '',
      phone: '',
      address: Address.empty(),
      shop: ShopDetail.empty(),
    );
  }
}

class Address {
  final String provinceTh;
  final String provinceEn;
  final String districtTh;
  final String districtEn;
  final String subDistrictTh;
  final String subDistrictEn;
  final String postCode;
  final String detail;

  Address({
    required this.provinceTh,
    required this.provinceEn,
    required this.districtTh,
    required this.districtEn,
    required this.subDistrictTh,
    required this.subDistrictEn,
    required this.postCode,
    required this.detail,
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    return Address(
      provinceTh: json['province_th'] ?? '',
      provinceEn: json['province_en'] ?? '',
      districtTh: json['district_th'] ?? '',
      districtEn: json['district_en'] ?? '',
      subDistrictTh: json['sub_district_th'] ?? '',
      subDistrictEn: json['sub_district_en'] ?? '',
      postCode: json['post_code'] ?? '',
      detail: json['detail'] ?? '',
    );
  }

  factory Address.empty() {
    return Address(
      provinceTh: '',
      provinceEn: '',
      districtTh: '',
      districtEn: '',
      subDistrictTh: '',
      subDistrictEn: '',
      postCode: '',
      detail: '',
    );
  }
}
